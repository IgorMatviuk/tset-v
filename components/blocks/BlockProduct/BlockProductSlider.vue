<template>
  <div class="product-slider">
    <div
      v-swiper:productSlider="swiperOptions"
      :cleanup-styles-on-destroy="false"
      class="product-slider__container"
    >
      <div class="swiper-wrapper product-slider__wrapper">
        <div
          v-for="(preview, index) in formattedPreviews"
          :key="index"
          class="swiper-slide product-slider__slide"
        >
          <BlockProductPreview ref="previewItem" :item="preview" />
        </div>
      </div>
      <SliderPagination class="product-slider__pagination" />
    </div>
  </div>
</template>

<script>
import BlockProductPreview from './BlockProductPreview'
import SliderPagination from '@/components/shared/SliderPagination'

import formattedPreviews from '@/mixins/formatted-previews'

export default {
  components: {
    BlockProductPreview,
    SliderPagination
  },
  mixins: [formattedPreviews],
  props: {
    previews: {
      type: Array,
      required: true
    }
  },
  data() {
    return {
      swiperOptions: {
        pagination: {
          el: '.product-slider__pagination',
          clickable: true
        },
        on: {
          slideChange: this.onSlideChange
        }
      }
    }
  },
  methods: {
    onSlideChange() {
      const video = this.$el.querySelector('video')

      if (this.productSlider.activeIndex === 1) {
        this.toggleVideoPlayState(video, 'play')
      } else {
        this.toggleVideoPlayState(video, 'pause')
      }
    },
    toggleVideoPlayState(video, nextState) {
      const isPlaying = !video.paused || !video.ended

      if (nextState === 'play') {
        video.play()
      } else if (isPlaying) {
        video.pause()
      }
    }
  }
}
</script>

<style lang="scss">
.product-slider {
  &__pagination {
    .swiper-pagination-bullet {
      bottom: 7px !important;
      width: 8px;
      height: 8px;
      margin: 0 5.5px !important;
    }
  }

  @include media-breakpoint-up(md) {
    display: none;

    &__pagination {
      display: none;
    }
  }
}
</style>
