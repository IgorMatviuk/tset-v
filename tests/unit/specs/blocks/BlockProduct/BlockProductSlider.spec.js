import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueAwesomeSwiper from 'vue-awesome-swiper'

import BlockProductSlider from '@/components/blocks/BlockProduct/BlockProductSlider'

import createFactory from '@/api/helpers/factory'
import productAttrs from '@/api/factories/product'

import '@/tests/unit/mocks/match-media'

const localVue = createLocalVue()
localVue.use(VueAwesomeSwiper)

describe('BlockProductSlider', () => {
  let previews

  beforeEach(() => {
    const products = createFactory(productAttrs, 12)
    const regularProduct = products.find((product) => product.in_stock)

    previews = regularProduct.previews
  })

  test('renders slider container', () => {
    const wrapper = shallowMount(BlockProductSlider, {
      localVue,
      propsData: {
        previews
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders slider container without video', () => {
    const products = createFactory(productAttrs, 12)
    const regularProduct = products.find((product) => product.in_stock)

    const wrapper = shallowMount(BlockProductSlider, {
      localVue,
      propsData: {
        previews: regularProduct.previews.filter(
          (item) => item.preview_type !== 'video'
        )
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
