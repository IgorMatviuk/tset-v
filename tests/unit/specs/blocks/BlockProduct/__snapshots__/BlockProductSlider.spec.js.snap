// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockProductSlider renders slider container 1`] = `
<div
  class="product-slider"
>
  <div
    class="product-slider__container swiper-container swiper-container-initialized swiper-container-horizontal"
  >
    <div
      class="swiper-wrapper product-slider__wrapper"
      style="webkit-transition-duration: 0ms; transition-duration: 0ms; webkit-transform: translate3d(NaNpx, 0px, 0px); transform: translate3d(NaNpx, 0px, 0px);"
    >
      <div
        class="swiper-slide product-slider__slide"
      >
        <blockproductpreview-stub
          item="[object Object]"
        />
      </div>
      <div
        class="swiper-slide product-slider__slide"
      >
        <blockproductpreview-stub
          item="[object Object]"
        />
      </div>
      <div
        class="swiper-slide product-slider__slide"
      >
        <blockproductpreview-stub
          item="[object Object]"
        />
      </div>
      <div
        class="swiper-slide product-slider__slide"
      >
        <blockproductpreview-stub
          item="[object Object]"
        />
      </div>
      <div
        class="swiper-slide product-slider__slide"
      >
        <blockproductpreview-stub
          item="[object Object]"
        />
      </div>
      <div
        class="swiper-slide product-slider__slide"
      >
        <blockproductpreview-stub
          item="[object Object]"
        />
      </div>
    </div>
     
    <sliderpagination-stub
      class="product-slider__pagination"
    />
    <span
      aria-atomic="true"
      aria-live="assertive"
      class="swiper-notification"
    />
  </div>
</div>
`;

exports[`BlockProductSlider renders slider container without video 1`] = `
<div
  class="product-slider"
>
  <div
    class="product-slider__container swiper-container swiper-container-initialized swiper-container-horizontal"
  >
    <div
      class="swiper-wrapper product-slider__wrapper"
      style="webkit-transition-duration: 0ms; transition-duration: 0ms; webkit-transform: translate3d(NaNpx, 0px, 0px); transform: translate3d(NaNpx, 0px, 0px);"
    >
      <div
        class="swiper-slide product-slider__slide"
      >
        <blockproductpreview-stub
          item="[object Object]"
        />
      </div>
      <div
        class="swiper-slide product-slider__slide"
      >
        <blockproductpreview-stub
          item="[object Object]"
        />
      </div>
      <div
        class="swiper-slide product-slider__slide"
      >
        <blockproductpreview-stub
          item="[object Object]"
        />
      </div>
      <div
        class="swiper-slide product-slider__slide"
      >
        <blockproductpreview-stub
          item="[object Object]"
        />
      </div>
      <div
        class="swiper-slide product-slider__slide"
      >
        <blockproductpreview-stub
          item="[object Object]"
        />
      </div>
    </div>
     
    <sliderpagination-stub
      class="product-slider__pagination"
    />
    <span
      aria-atomic="true"
      aria-live="assertive"
      class="swiper-notification"
    />
  </div>
</div>
`;
